### PR TITLE
[DLPack] Add support for missing keyword-arguments.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150691
* __->__ #150218
* #150217
* #150216

This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

### CODE REVIEW

File: aten/src/ATen/DLConvertor.cpp [hunk 1]
```diff
@@ -388,4 +389,35 @@ Tensor fromDLPackVersioned(DLManagedTensorVersioned* src, std::function<void(voi
...
+        static_cast<c10::DeviceIndex>(optional_dl_device->device_id));
+
+    if (device != data.device()) {
+      TORCH_CHECK(
+          !force_move,
+          "cannot move tensor from ",
+          data.device(),
+          " to ",
+          device,
+          " without copying. Set copy=True is needed.");
```
Comment (msaroufim) [inline aten/src/ATen/DLConvertor.cpp:411]: just double checking this check would make it clear that users would need to set the copy flag in

`torch.from_dlpack(..., copy=copy)`

File: torch/utils/dlpack.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ class DLDeviceType(enum.IntEnum):
 
 # TODO: add a typing.Protocol to be able to tell Mypy that only objects with
 # __dlpack__ and __dlpack_device__ methods are accepted.
-def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
+def from_dlpack(ext_tensor: Any, device: Any = None, copy: Optional[bool] = None) -> 'torch.Tensor':
```
Comment (albanD) [inline torch/utils/dlpack.py:57]: kwarg-only!

Comment (albanD) [inline torch/utils/dlpack.py:57]: Docstring needs updating.
Also is the "device" type really any? We have  Device in our torch.types utils if needed.

File: torch/csrc/Module.cpp [hunk 3]
```diff
@@ -636,6 +668,22 @@ static PyObject* THPModule_fromDLPack(PyObject* _unused, PyObject* data) {
...
+    PyObject* _unused,
+    PyObject* data) {
+  HANDLE_TH_ERRORS
+  TORCH_CHECK(
+      THPDevice_Check(data),
+      "torchDeviceToDLDevice: expected torch.device argument.");
+  auto device = reinterpret_cast<THPDevice*>(data)->device;
+  auto dl_device = at::torchDeviceToDLDevice(device);
+  auto tuple = PyTuple_New(2);
+  PyTuple_SetItem(tuple, 0, THPUtils_packInt64(dl_device.device_type));
```
Comment (albanD) [inline torch/csrc/Module.cpp:681]: nit here and below for perf
```suggestion
  PyTuple_SET_ITEM(tuple, 0, THPUtils_packInt64(dl_device.device_type));
```

File: torch/csrc/Module.cpp [hunk 4]
```diff
@@ -636,6 +668,22 @@ static PyObject* THPModule_fromDLPack(PyObject* _unused, PyObject* data) {
...
+static PyObject* THPModule_torchDeviceToDLDevice(
+    PyObject* _unused,
+    PyObject* data) {
+  HANDLE_TH_ERRORS
+  TORCH_CHECK(
+      THPDevice_Check(data),
+      "torchDeviceToDLDevice: expected torch.device argument.");
+  auto device = reinterpret_cast<THPDevice*>(data)->device;
+  auto dl_device = at::torchDeviceToDLDevice(device);
+  auto tuple = PyTuple_New(2);
```
Comment (albanD) [inline torch/csrc/Module.cpp:680]: Add error checking in case this failed

File: aten/src/ATen/DLConvertor.cpp [hunk 5]
```diff
@@ -96,10 +96,14 @@ DLDataType getDLDataType(const Tensor& t) {
   return dtype;
 }
 
-static DLDevice getDLDevice(const Tensor& tensor, c10::DeviceIndex device_id) {
+DLDevice torchDeviceToDLDevice(at::Device device) {
   DLDevice ctx;
-  ctx.device_id = static_cast<int32_t>(static_cast<unsigned char>(device_id));
-  switch (tensor.device().type()) {
+
+  ctx.device_id = (device.is_cuda() || device.is_privateuseone())
```
Comment (albanD) [inline aten/src/ATen/DLConvertor.cpp:102]: Why this new logic? Why would we ignore the device index being passed in?

File: aten/src/ATen/DLConvertor.cpp [hunk 6]
```diff
@@ -130,38 +133,40 @@ static DLDevice getDLDevice(const Tensor& tensor, c10::DeviceIndex device_id) {
...
 #ifdef USE_ROCM
       // this looks funny, we need to return CUDA here to masquerade
-      return at::Device(DeviceType::CUDA, static_cast<c10::DeviceIndex>(ctx.device_id));
+      return at::Device(DeviceType::CUDA, index);
 #else
-      return at::Device(DeviceType::HIP, static_cast<c10::DeviceIndex>(ctx.device_id));
+      return at::Device(DeviceType::HIP, index);
 #endif
     case DLDeviceType::kDLOneAPI:
+      TORCH_CHECK(data != nullptr, "Can't get ATen device for XPU without XPU data.");
```
Comment (albanD) [inline aten/src/ATen/DLConvertor.cpp:161]: nit: Why do we use data here and not the passed in type and index?

File: aten/src/ATen/DLConvertor.cpp (hunk 1 above)
Comment (albanD) [inline aten/src/ATen/DLConvertor.cpp:411]: copy=None would also work. It's more that you cannot set copy=False in this case

File: aten/src/ATen/DLConvertor.cpp [hunk 7]
```diff
@@ -388,4 +389,35 @@ Tensor fromDLPackVersioned(DLManagedTensorVersioned* src, std::function<void(voi
...
+          data.device(),
+          " to ",
+          device,
+          " without copying. Set copy=True is needed.");
+      return data.to(device);
+    }
+  }
+
+  if (force_copy) {
+    return data.detach().clone();
```
Comment (albanD) [inline aten/src/ATen/DLConvertor.cpp:417]: There should be no autograd here already right? So detach() is not needed

File: torch/_tensor.py [hunk 8]
```diff
@@ -1673,7 +1673,13 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
 
     __torch_dispatch__ = _C._disabled_torch_dispatch_impl
 
-    def __dlpack__(self, stream=None, max_version=None):
+    def __dlpack__(
+        self,
+        stream: Optional[Any] = None,
```
Comment (albanD) [inline torch/_tensor.py:1678]: kwargs only

File: torch/_tensor.py [hunk 9]
```diff
@@ -1673,7 +1673,13 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
...
+        max_version: Optional[tuple[int, int]] = None,
+        dl_device: Optional[tuple[enum.IntEnum, int]] = None,
+        copy: Optional[bool] = None,
```
Comment (albanD) [inline torch/_tensor.py:1681]: docstring update

File: torch/csrc/Module.cpp [hunk 10]
```diff
@@ -607,25 +607,57 @@ void DLPack_Capsule_Destructor(PyObject* data) {
...
+          static_cast<int32_t>(dl_device[1])};
+    }
+
+    auto tensor = at::DLPackTraits<T>::toDLPack(
+        at::maybeCopyTensor(data, optional_dl_device, copy));
+    return PyCapsule_New(
+        tensor, at::DLPackTraits<T>::capsule, DLPack_Capsule_Destructor<T>);
+  }
+
+  return nullptr;
```
Comment (albanD) [inline torch/csrc/Module.cpp:643]: I'm don't recall, will this have the right python err set?

File: torch/utils/dlpack.py [hunk 11]
```diff
@@ -107,19 +107,34 @@ def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
...
+                device = torch.device(device)
+            assert isinstance(device, torch.device), (
+                f"from_dlpack: unsupported device type: {type(device)}"
+            )
+            device = torch._C._torchDeviceToDLDevice(device)
+
+        kwargs["dl_device"] = device
+
+        ext_device = ext_tensor.__dlpack_device__()
+        # ext_device is either CUDA or ROCm, we need to pass the current
```
Comment (albanD) [inline torch/utils/dlpack.py:128]: Note that a lot of this string handling should get generalized to any device which is an accelerator in PT.
Ok to do later, but as we're going to see issues with hip or xpu, we should just refactor it together

File: torch/utils/dlpack.py [hunk 12]
```diff
@@ -129,11 +144,18 @@ def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
             # Try running __dlpack__ while specifying `max_version` argument.
             dlpack = ext_tensor.__dlpack__(**kwargs)
         except TypeError:
-            # If that doesn't work, try removing the `max_version` argument.
+            # If that doesn't work, try removing the all the arguments but those
+            # corresponding to DLPack 2022.12 specification.
             kwargs.pop("max_version")
+            kwargs.pop("dl_device")
+            kwargs.pop("copy")
```
Comment (albanD) [inline torch/utils/dlpack.py:151]: Is it actually ok to ignore these arguments?!

File: aten/src/ATen/DLConvertor.cpp (hunk 5 above)
Comment (ysiraichi) [reply aten/src/ATen/DLConvertor.cpp:102]: I just got it from the place it was being called before. Needed it, since I'm also calling this function outside.

File: aten/src/ATen/DLConvertor.cpp (hunk 6 above)
Comment (ysiraichi) [reply aten/src/ATen/DLConvertor.cpp:161]: I'm not really sure myself. I just added this check because I needed to use this function without `data` when copying the tensor to some other device (see `maybeCopyTensor`).

File: torch/csrc/Module.cpp (hunk 10 above)
Comment (ysiraichi) [reply torch/csrc/Module.cpp:643]: I think so. At least, it will throw an error inside the parser. I will replace it with `Py_RETURN_NONE`, just for consistency.

File: torch/csrc/Module.cpp (hunk 4 above)
Comment (ysiraichi) [reply torch/csrc/Module.cpp:680]: Not sure I get what you mean. If `at::torchDeviceToDLDevice` errors, the error will bubble up to Python, won't it?

File: torch/utils/dlpack.py (hunk 11 above)
Comment (ysiraichi) [reply torch/utils/dlpack.py:128]: Could you help me understand what string handling you are referring to? Doesn't `torch.device(device)` work with any accelerator in PT?

File: torch/csrc/Module.cpp (hunk 10 above)
Comment (ysiraichi) [reply torch/csrc/Module.cpp]: I think I hadn't understood you question earlier. If you are asking whether a C++ exception will be mapped the correct Python error set, the answer is: yes! `END_HANDLE_TH_ERRORS` will take care of that. Specifically, `torch::translate_exception_to_python(std::current_exception())` call will do the job.

File: torch/csrc/Module.cpp (hunk 4 above)
Comment (albanD) [reply torch/csrc/Module.cpp:680]: CPython is C API, not C++. There is no error throwing here, you need to check the tuple is not null and if so, make this a c++ error

```cpp
if (!tuple) {
  throw python_error();
}
```

File: aten/src/ATen/DLConvertor.cpp (hunk 5 above)
Comment (albanD) [reply aten/src/ATen/DLConvertor.cpp:102]: Not sure what you mean by this?

File: aten/src/ATen/DLConvertor.cpp (hunk 6 above)
Comment (albanD) [reply aten/src/ATen/DLConvertor.cpp:161]: cc @EikanWang do you know by any chance?

File: torch/csrc/Module.cpp (hunk 10 above)
Comment (albanD) [reply torch/csrc/Module.cpp]: Returning None here is completely different. These are in no way interchangeable.
Returning `nullptr` from these APIs mean that something went wrong and the caller should check the globally set error for more info. Returning None means that all went well and the result is "None".
You either want one or the other :D 
Or maybe you're saying this is dead code?

File: torch/utils/dlpack.py (hunk 11 above)
Comment (albanD) [reply torch/utils/dlpack.py:142]: Only the fact that you do custom processing only for CUDA and ROCm but we have quite a few other accelerators that would require the same treatment in the future.

Comment (ysiraichi) [reply torch/utils/dlpack.py:142]: Since I'm not familiar with these other accelerators, I think I will leave it for a future PR. In this one, I'm simply adding support for the extra keywords (in this case, passing them through to `__dlpack__()` method). What do you think?

File: aten/src/ATen/DLConvertor.cpp (hunk 5 above)
Comment (ysiraichi) [reply aten/src/ATen/DLConvertor.cpp:102]: Previously, this is how we were calling this function (the only call site):

https://github.com/pytorch/pytorch/blob/e1f28fe17bc55389fc5a31bad588d22e5ecca722/aten/src/ATen/DLConvertor.cpp#L317-L321

So, I basically, moved that code into this function. In other words, the logic did not change. The device index will always be either `src.get_device()` (i.e. `src.device().index()`), if the condition `src.is_cuda() || src.is_privateuseone()` is true, or 0 otherwise.

File: torch/csrc/Module.cpp (hunk 10 above)
Comment (ysiraichi) [reply torch/csrc/Module.cpp]: This is essentially deadcode.
On second thoughts, it would be better to just `TORCH_INTERNAL_ASSERT(r.idx == 0);`. Then, there would be no need to return None, since the bug would be in the arg parser.

File: torch/utils/dlpack.py (hunk 11 above)
Comment (albanD) [reply torch/utils/dlpack.py:142]: Yes let's just open an issue for it and add the xpu and privateuse1 labels on it.

File: aten/src/ATen/DLConvertor.cpp (hunk 5 above)
Comment (albanD) [reply aten/src/ATen/DLConvertor.cpp:102]: Ok, I guess it is another case where we should have an issue and tag other accelerator backends.

**Comment by ysiraichi** [comment]
Summarizing the current status of this PR:

1. We are currently waiting for @EikanWang feedback on the reason we need the data pointer in order to get a torch device
2. TODO: open an issue for XPU and PrivateUse1, regarding stream handling in DLPack
3. TODO: open an issue, regarding the way we translate torch device into DLPack device

In my opinion, since (I think) I'm not really changing the behavior of these functions (just slightly refactoring them), we could land these changes and open issues as needed.

@albanD do you think it's better for us to wait for (1) before landing this PR? 
Let me know what you think.

File: aten/src/ATen/DLConvertor.cpp [hunk 13]
```diff
@@ -388,4 +389,35 @@ Tensor fromDLPackVersioned(DLManagedTensorVersioned* src, std::function<void(voi
...
+    const Tensor& data,
+    std::optional<DLDevice> optional_dl_device,
+    std::optional<bool> copy) {
+  bool force_copy = copy.has_value() && *copy;
+  bool force_move = copy.has_value() && !*copy;
+
+  if (optional_dl_device.has_value()) {
+    auto device = at::getATenDevice(
+        optional_dl_device->device_type,
+        static_cast<c10::DeviceIndex>(optional_dl_device->device_id));
```
Comment (albanD) [inline aten/src/ATen/DLConvertor.cpp:402]: Shouldn't you pass data in for xpu to work?

**Review by albanD — APPROVED** [review]
I'm ok with skipping XPU handling until we hear back from them.
It might be a bit broken though, not sure if we have a way to test it?

File: aten/src/ATen/DLConvertor.cpp (hunk 6 above)
Comment (guangyey) [reply aten/src/ATen/DLConvertor.cpp:161]: Currently, XPU always gets its device index from data since device index in PyTorch might not match the order of the xpu driver enumerated.
We will refine this issue in the following PR when the appropriate xpu toolchain is available.
BTW, here `TORCH_CHECK` is not needed since `getDeviceFromPtr` will check data's validity inside itself.

File: aten/src/ATen/DLConvertor.cpp (hunk 13 above)
Comment (guangyey) [reply aten/src/ATen/DLConvertor.cpp:402]: Yes, currently, XPU requires access to `data` within `getATenDevice`. @gujinghui could we refine this logic?

Comment (gujinghui) [reply aten/src/ATen/DLConvertor.cpp:402]: Can we pass the data ptr here to not break xpu? We still need to check the device id upon the data ptr for a long term. Thanks.

Comment (ysiraichi) [reply aten/src/ATen/DLConvertor.cpp:402]: Is this the correct thing to do, here, though? The problem I see is that the `const Tensor& data` parameter in this context does not necessarily live in the XPU device. In summary:

- **`data` is a tensor that, if necessary (e.g. `copy=true`), must be copied to `device = at::getATenDevice()`**
- `optional_dl_device` is something the user specifies when calling `tensor.__dlpack__()` (likely from a `otherlib.from_dlpack(tensor, device = resultShouldLiveHere)`)

So, I guess, the actual problem here is: how do we retrieve a torch device for XPU? Or, what device should we pass to `tensor.to()` in order to move a tensor to XPU.

-----

My guess is that the solution here is **to branch** on XPU device:

```c++
c10::Device device;
if (optional_dl_device->device_type == DLDeviceType::kDLOneAPI) {
    device = torch.device(c10::kXPU);
} else {
    device = at::getATenDevice(...);
}
```

The problem with this is that **we will be ignoring the device index**, which I'm not sure is ideal. Is there a way to create a tensor on a specific XPU device (accounting for the given index)?

@guangyey @gujinghui
What do you think?

**Comment by ysiraichi** [comment]
@guangyey @gujinghui @albanD
After adding support for moving tensors onto XPU, I believe we need tests for it. So, I'm leaving it for a future PR.
I will probably try to merge this stack next Wednesday. So, if you think I should wait and fix this first, let me know until then.

**Comment by albanD** [comment]
Happy to leave XPU as-is for now and cleanup later yes.
